// @uiw/react-markdown-preview/esm/index.js
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import _extends from "@babel/runtime/helpers/extends";
var _excluded = ["prefixCls", "className", "source", "style", "onScroll", "onMouseOver", "pluginsFilter", "warpperElement"];
import React, { useImperativeHandle } from 'react';
import ReactMarkdown from 'react-markdown';
import gfm from 'remark-gfm';
import slug from 'rehype-slug';
import headings from 'rehype-autolink-headings';
import rehypeRaw from 'rehype-raw';
import rehypeAttrs from 'rehype-attr';
import rehypePrism from 'rehype-prism-plus';
import rehypeRewrite, { getCodeString } from 'rehype-rewrite';
import { octiconLink } from './nodes/octiconLink';
import { copyElement } from './nodes/copy';
import "./styles/markdown.css";
import { jsx as _jsx } from "react/jsx-runtime";

var rehypeRewriteHandle = (node, index, parent) => {
  if (node.type === 'element' && parent && parent.type === 'root' && /h(1|2|3|4|5|6)/.test(node.tagName)) {
    var child = node.children && node.children[0];

    if (child && child.properties && child.properties.ariaHidden === 'true') {
      child.properties = _extends({
        class: 'anchor'
      }, child.properties);
      child.children = [octiconLink];
    }
  }

  if (node.type === 'element' && node.tagName === 'pre') {
    var code = getCodeString(node.children);
    node.children.push(copyElement(code));
  }
};

export default /*#__PURE__*/React.forwardRef((props, ref) => {
  var {
    prefixCls = 'wmde-markdown wmde-markdown-color',
    className,
    source,
    style,
    onScroll,
    onMouseOver,
    pluginsFilter,
    warpperElement = {}
  } = props,
      other = _objectWithoutPropertiesLoose(props, _excluded);

  var mdp = /*#__PURE__*/React.createRef();
  useImperativeHandle(ref, () => _extends({}, props, {
    mdp
  }), [mdp, props]);
  var cls = (prefixCls || '') + " " + (className || '');
  var rehypePlugins = [[rehypePrism, {
    ignoreMissing: true
  }], rehypeRaw, slug, headings, [rehypeRewrite, {
    rewrite: rehypeRewriteHandle
  }], [rehypeAttrs, {
    properties: 'attr'
  }], ...(other.rehypePlugins || [])];
  var remarkPlugins = [...(other.remarkPlugins || []), gfm];
  return /*#__PURE__*/_jsx("div", _extends({
    ref: mdp,
    onScroll: onScroll,
    onMouseOver: onMouseOver
  }, warpperElement, {
    className: cls,
    style: style,
    children: /*#__PURE__*/_jsx(ReactMarkdown, _extends({}, other, {
      rehypePlugins: pluginsFilter ? pluginsFilter('rehype', rehypePlugins) : rehypePlugins,
      remarkPlugins: pluginsFilter ? pluginsFilter('remark', remarkPlugins) : remarkPlugins,
      children: source || ''
    }))
  }));
});
